### PR TITLE
tar build: support uploading local source

### PR DESCRIPTION
**Pull request by northtyphoon**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [ ] Each command and parameter has a meaningful description.
- [ ] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 1]
```diff
@@ -83,4 +83,12 @@ def load_arguments(self, _):
...
+        c.argument('build_id', help='The unique build identifier.')
+        c.argument('registry_name', options_list=['--name', '-n'], help='The name of the container registry. You can configure the default registry name using `az configure --defaults acr=<registry name>`', completer=get_resource_name_completion_list(REGISTRY_RESOURCE_TYPE), configured_default='acr')
+        c.argument('image_name', options_list=['--tag', '-t'], help="The image name and optionally a tag in the 'name:tag' format.")
+        c.argument('docker_file_path', options_list=['--file', '-f'], help="The relative path of the the docker file to the source code root folder (Default is 'Dockerfile').")
+        c.argument('timeout', help='The build timeout in seconds.')
+        c.argument('build_args', nargs='+', help='The space-separated build arguments in a format of <name>=<value>.')
+        c.argument('secret_build_args', nargs='+', help='The space-separated secret build arguments in a format of <name>=<value>.')
+        c.argument('show_logs', nargs='?', required=False, const='true', default='true', help='Indicates whether showing logs after succesuflly queuing the build.', choices=['true', 'false'])
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: I'd assume that it needs to be `default=True` and can `const` be deleted [from [argparse](https://docs.python.org/3/library/argparse.html#const)]? It might be required depending on how the CLI invokes the lib; did you test the default? Maybe the lib is smart enough to parse a string into a bool later.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 2]
```diff
@@ -83,4 +83,12 @@ def load_arguments(self, _):
         c.argument('replication_name', help='The name of the replication. Default to the location name.', completer=None)
 
     with self.argument_context('acr build') as c:
-        c.argument('build_id', help='The unique build identifier.')
\ No newline at end of file
+        c.argument('build_id', help='The unique build identifier.')
+        c.argument('registry_name', options_list=['--name', '-n'], help='The name of the container registry. You can configure the default registry name using `az configure --defaults acr=<registry name>`', completer=get_resource_name_completion_list(REGISTRY_RESOURCE_TYPE), configured_default='acr')
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: You don't need to redefine this unless you want to use a different definition from what's defined [here ](https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py#L29)

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/sdk/models/azure-mgmt-containerregistry_enums.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+class OsType(Enum):
+
+    windows = "Windows"
+    linux = "Linux"
+
+
+class BuildStatus(Enum):
+
+    queued = "Queued"
+    started = "Started"
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/sdk/models/azure-mgmt-containerregistry_enums.py:46]: what's the difference between Started and Running?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 1 above)
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Check the spellings ...

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 4]
```diff
@@ -83,4 +83,12 @@ def load_arguments(self, _):
...
+        c.argument('image_name', options_list=['--tag', '-t'], help="The image name and optionally a tag in the 'name:tag' format.")
+        c.argument('docker_file_path', options_list=['--file', '-f'], help="The relative path of the the docker file to the source code root folder (Default is 'Dockerfile').")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: `the the`. Please fix help messages globally and reuse the ones from swagger if possible.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 5]
```diff
@@ -1,78 +1,109 @@
...
-from knack.util import CLIError
 from ._utils import (
-    get_resource_group_name_by_registry_name,
+    get_resource_group_name_by_registry_name
 )
-from azure.cli.core.commands import LongRunningOperation
-import pytz
+from azure.cli.core.commands import (
+    LongRunningOperation
+)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: `from azure.cli.core.commands import LongRunningOperation` if you are only importing one.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 6]
```diff
@@ -1,78 +1,109 @@
...
+from knack.util import CLIError
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class BlobInfo(object):
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: The class seems to be an overkill since each field is passed separately.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 7]
```diff
@@ -1,78 +1,109 @@
...
+
     resource_group_name = get_resource_group_name_by_registry_name(
         cmd.cli_ctx, registry_name, resource_group_name)
 
     # Required and hardcoded for now. Refer to the "HACK" introduced in build_operations.py
     custom_headers = dict()
     custom_headers['logType'] = 'RawText'
     build_log_result = client.get_log_link(
-        build_id=build_id, resource_group_name=resource_group_name,
+        build_id=build_id, resource_group_name=resource_group_name,
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: NIT: Remove this trailing space. Run pylint at one point to fix such issues.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 8]
```diff
@@ -1,78 +1,109 @@
...
+                AppendBlobService(account_name=blob_info.account_name, sas_token=blob_info.sas_token, endpoint_suffix=blob_info.endpoint_suffix),
+                blob_info.container_name, blob_info.blob_name)
 
 
-def __get_match(sas_url):
+def _get_match(sasURL):
     return re.search(
-        (r"http(s)?://(?P<account_name>.*?)\..*?/(?P<container_name>.*?)/"
-            r"(?P<blob_name>.*?)\?(?P<sas_token>.*)"), sas_url)
+        (r"http(s)?://(?P<account_name>.*?)\.blob\.(?P<endpoint_suffix>.*?)/(?P<container_name>.*?)/"
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Just curious: when's the case where `endpoint_suffix` is required (it defaults to `core.windows.net`)?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 9]
```diff
@@ -1,78 +1,109 @@
...
-    __stream_logs(byte_size, timeout_in_seconds, blob_service,
-                container_name, blob_name)
+
+    _stream_logs(byte_size, timeout_in_seconds, 
+                AppendBlobService(account_name=blob_info.account_name, sas_token=blob_info.sas_token, endpoint_suffix=blob_info.endpoint_suffix),
+                blob_info.container_name, blob_info.blob_name)
 
 
-def __get_match(sas_url):
+def _get_match(sasURL):
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: `sas_url`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 10]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
         cmd.cli_ctx, registry_name, resource_group_name)
-    build_parameters = DockerBuildParameters(docker_file_path)
 
-    if context_path:
-        build_parameters.context_path = context_path
+    if docker_file_path is None:
+        docker_file_path = "Dockerfile"
+
+    build_parameters = DockerBuildParameters(docker_file_path)
+    # context_path in tar build is alwasy the source code root folder
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: ultra-nit: typo on "always"

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 11]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+    else:
+        raise CLIError("'{}' doesn't exist.".format(source_location))
+
+
+def _check_image_name(image_name):
+    
+    # referenc: https://github.com/docker/distribution/tree/master/reference
+
+    if image_name == "":
+        raise CLIError("'--tag' value should not be empty.")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Why are you checking `image_name` but error on `--tag`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 12]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+        return source_location
+    else:
+        raise CLIError("'{}' doesn't exist.".format(source_location))
+
+
+def _check_image_name(image_name):
+    
+    # referenc: https://github.com/docker/distribution/tree/master/reference
+
+    if image_name == "":
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: `if not image_name`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 13]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+        raise CLIError("'--tag' value should not be empty.")
+    
+    tokens = image_name.split(':')
+    if(len(tokens) > 2):
+        raise CLIError("'--tag' value should be repository-name and optionally a tag in the 'repository-name:tag' format")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: NIT: our convention has been to just use `repository:tag`.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 14]
```diff
@@ -83,4 +83,12 @@ def load_arguments(self, _):
         c.argument('replication_name', help='The name of the replication. Default to the location name.', completer=None)
 
     with self.argument_context('acr build') as c:
-        c.argument('build_id', help='The unique build identifier.')
\ No newline at end of file
+        c.argument('build_id', help='The unique build identifier.')
+        c.argument('registry_name', options_list=['--name', '-n'], help='The name of the container registry. You can configure the default registry name using `az configure --defaults acr=<registry name>`', completer=get_resource_name_completion_list(REGISTRY_RESOURCE_TYPE), configured_default='acr')
+        c.argument('image_name', options_list=['--tag', '-t'], help="The image name and optionally a tag in the 'name:tag' format.")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: `--tag` is already used to refer to tag (e.g., latest). `image_name` by definition is `repository:tag`. `-t` is tenant, should avoid using it on a different parameter.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 15]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+
+    # check repository-name    
+    repository_name = tokens[0]
+    if len(repository_name) > 255:
+        raise CLIError("The repository-name of '--tag' value should be no more than 255 characters.")
     else:
-        # Maybe just raise error?
-        return result
+        if re.match(r"^[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?$", repository_name) is None:
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: IMO an overkill. Client app should avoid validation. Let the server side validate.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 16]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+            print("Starting to stream the logs...")
+            return acr_build_show_logs(cmd, client, registry_name, result.build_id, resource_group_name)
+    except Exception as err:
+        raise CLIError(err)
+
+
+def _check_remote_source_code(source_location):
+
+    # NOTE: acr_build doesn't support git repo url
+    if source_location.lower().startswith("git://"):
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: I think we probably want to support the others too (`git@`, `github.com/`). For example you should be able to do `git@bitbucket.org:atlassianlabs/atlassian-docker.git` which would be different than github - same thing for lots of other providers. Don't have to tackle this now, we can just add a work item.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 17]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+
+def _upload_source_code(client, registry_name, resource_group_name, source_location):
+
+    tar_file_path = os.path.join(tempfile.gettempdir(),
+                                 "source_archive_{}.tar.gz".format(hash(os.times())))
+
+    try:
+        logger.debug("Starting to archive the source code to '{}'.".format(tar_file_path))
+
+        # TODO: Check if it works in windows
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: You can use `platform.system() == 'Windows'` to find our what the right separator is.

**Review by djyou — CHANGES_REQUESTED** [review]

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 15 above)
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Maybe a TODO in VSTS for this, we should pull in their tests and compare the differences.  Did you pull this from the Python docker cli?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 18]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+        if not source_location.endswith('/'):
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: remove this; just use `os.path.join(path, next-thing)` for all the below calls, it will do the right thing. or you can do `os.path.sep`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 15 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Just avoid client side validation and let the server side validate. You'll run into maintenance and inconsistency headache easily with client validation.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 19]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+
+            logger.debug(".dockerignore: no rule for '{}'.".format(tarinfo.name))
+            return tarinfo
+
+        with tarfile.open(tar_file_path, "w:gz") as tar:
+            tar.add(source_location, arcname=os.path.basename(source_location), filter=_filter_file)
+
+        logger.debug("Starting to upload the archived source code from '{}'.".format(tar_file_path))
+
+        source_upload_location = client.get_source_upload_url(
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: we should call this one before making the tar, it would be really sad to have a transient error here after spending a few mins making the tar - the block blob service should have retries automatically i believe, worth double checking

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 20]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+            else:
+                tokens[index] = token.replace("*", "[^/]*").replace("?", "[^/]")
+
+        self.pattern = "^{}$".format("/".join(tokens))
+
+
+def _load_dockerignore_file(source_location):
+
+    # reference: https://docs.docker.com/engine/reference/builder/#dockerignore-file
+    dokcer_ignore_file = os.path.join(source_location, ".dockerignore")
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: nit: typo

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 15 above)
Comment (ehotinger) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: @djyou this isn't really something a server can validate...

Comment (ehotinger) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: @djyou idk, for most of these I think it makes sense to have the client-side validation. For this one, it would be a bummer to send the request all the way down the entire pipeline just to have it fail, pretty bad user experience; we'd bill them for it; we'd hurt our resources. For the others we pretty much have to do them.

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: A preferred pattern is to avoid client side validation because if tomorrow there is a change on the server side we will have to update the client side too. The even worse part is we have to ask users to update their client to match whatever the new server side validation does.

I suppose server side would have to validate image name anyway. We probably have more traffic from various SDKs directly or third party apps than from our own clients (i.e., CLI/Portal), plus there is no way we can prevent incoming garbage data. I agree with your concerns but compared with the potential maintenance effort, the real overhead on the user experience is no more than a few seconds. We can start billing after all validations are passed to avoid the billing issue.

That said, in some rare case client side validation can make a real difference so it's just a trade-off you folks can decide.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/sdk/models/azure-mgmt-containerregistry_enums.py (hunk 3 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/sdk/models/azure-mgmt-containerregistry_enums.py:46]: I discussed with @mnltejaswini the other day and consider to remove Started.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 2 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Oh, I migrated the code from the extension which required the definition. I will remove it.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 6 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Not sure if I understand. The class takes a full sas url and break it down to account/container/endpoint/sas.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 12 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: fixed

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 13 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: fixed

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 14 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: For this particular command, we want to follow docker build convention.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 20 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: fixed

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 18 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: The purpose of the code is to append / so that tarfile won't add the passed-in source-location to the name which sometimes can block unpacking. I need to do some tests on windows to check the solution.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 10 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: fixed

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 1 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: I see acr create command has both default and const. @djyou, do you know why?

Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: I tested. If I removed const='true', --show-logs wouldn't be set. I think I need to keep it.

Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: fixed

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 4 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: fixed

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 5 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: fixed. I always suggest people add {} or () at the beginning because it's very likely we will add more and {} also makes code easy to read (IMO).

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 19 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: updated.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 6 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: You can have a method that returns all four of them. The reason is that I didn't think the object is passed around anywhere so seems to be an overkill. The current approach is fine too.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 14 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Not sure about this one. `--tag` definitely conflicts with existing parameters. Does docker `-t` mean `--tag` or `--image-name` or something else? You might be able to use `-t` since although it's generally tenant it's not used anywhere under `acr` group.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 21]
```diff
@@ -1,78 +1,109 @@
...
+        (r"http(s)?://(?P<account_name>.*?)\.blob\.(?P<endpoint_suffix>.*?)/(?P<container_name>.*?)/"
+            r"(?P<blob_name>.*?)\?(?P<sas_token>.*)"), sasURL)
 
 
-def __stream_logs(byte_size,
-                timeout_in_seconds,
-                blob_service,
-                container_name,
-                blob_name):
+def _stream_logs(byte_size,
```
Comment (sajayantony) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:76]: >_stream_logs [](start = 4, length = 12)

I wonder if we can move all log streaming operations to its own file.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 22]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
     return True
 
 
-def acr_build_queue(cmd,
+def acr_queue(cmd,
```
Comment (sajayantony) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: >acr_queue [](start = 4, length = 9)

Do we need to prefix this with acr or can we just call it queue_build?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 23]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
     is_push_enabled = True
     if image_name is None:
         is_push_enabled = False
+        print("'--tag' is not providied. Skip image push after build.")
+    else:
+        image_name = _check_image_name(image_name)
 
     # hard-code platform to linux and cpu to 1
     platform = PlatformProperties("Linux")
     platform.cpu = 1
```
Comment (sajayantony) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:243]: >cpu = 1 [](start = 13, length = 7)

This should default to 2 core. Also the value of 1 should be failing on the RP.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 24]
```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
...
+            platform=platform,
+            build_arguments=build_arguments)
+
+        result = LongRunningOperation(cmd.cli_ctx)(client.queue(
+            build_request=build_request, resource_group_name=resource_group_name, registry_name=registry_name))
+
+        print("Successfully queued a build with build-id: {}.".format(result.build_id))
+
+        if show_logs == 'true':
+            print("Starting to stream the logs...")
```
Comment (sajayantony) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: >print("Starting to stream the logs...") [](start = 12, length = 39)

We should reword this to indicate something like  "Build {id} initiated." Log streaming is expected here and so we don't have to explicitly call it out. 
Maybe run a timer that outputs waiting for agent if we don't see the logs in say 3-5 seconds.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 16 above)
Comment (sajayantony) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: At this point I think we can assume anything that isn't a sas url pattern is a git repo. 

---
In reply to: [168360092](https://github.com/AppPlatPerf/azure-cli/pull/60#discussion_r168360092) [](ancestors = 168360092)

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 15 above)
Comment (sajayantony) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: I'm ok with this since we can avoid a round trip and the image name regex is more of a spec. 

---
In reply to: [168359952](https://github.com/AppPlatPerf/azure-cli/pull/60#discussion_r168359952) [](ancestors = 168359952)

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py (hunk 14 above)
Comment (sajayantony) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: We are aligning with docker on this. so -t is ok.

---
In reply to: [168378359](https://github.com/AppPlatPerf/azure-cli/pull/60#discussion_r168378359) [](ancestors = 168378359)

**Review by sajayantony — APPROVED** [review]
:shipit:

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 24 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Build id is output after successfully queuing the build. 

The command takes an option to disable show-logs. So this message is output separately.

We can add a progress indicator or something in ```_stream_log``` method.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 17 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: I verified. "/" works on both windows and linux.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 22 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: The command will be moved to acr command group once we integrate the official sdk.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 23 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:243]: We currently default to 1. Why 2?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 21 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:76]: @ehotinger what do you think?

Comment (ehotinger) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:76]: I'm fine with it either way. I think most of the other command sections are all self-contained, it does make it easy to find stuff. On the other hand, our file has already gotten quite big and we still need to add more. If we do this, maybe we should do the same for all other commands as well. E.g., all the tar packaging code here as well. Maybe we can have `log_stream_utils.py` or something similar. `acr_show_logs` or whatever the command is called now still has to remain here though.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 22 above)
Comment (ehotinger) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: I don't think we have to use the prefix, maybe they were sticklers for having some prefix in the past. I see a lot of teams that don't follow the convention, e.g., cosmosdb. At the very least for now it keeps us consistent with the rest of the work @djyou has done and the only one it affects is the one reading the source code of the CLI.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 18 above)
Comment (ehotinger) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: @northtyphoon I see, I was looking at the dockerignore case and not the issue with the tar itself. You should just be able to use `os.path.sep` here and it will handle all the cases. I'm guessing both of these work on Windows 10, but we should use the os independent versions anyways.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 22 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: It should actually be `acr_build` as the command is `acr build`. Simple way just replace space with underscore from command as method name.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 23 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:243]: Shouldn't this be a swagger level default value? Add to swagger if it's not there and don't set it in CLI.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 24 above)
Comment (sajayantony) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: You are right the output does should the build id before streaming the logs - 
```
Successfully queued a build with build-id: f32f514dfe1e493080457384a05d7f1d.
Starting to stream the logs...
time="2018-02-15T01:48:36Z" level=info msg="Running command docker login -u 00000000-0000-0000-0000-000000000000 -p *********
```

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 23 above)
Comment (sajayantony) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:243]: I believe you allocate 2 cores by default. /cc @mnltejaswini do we validate this on the RP?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 15 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Also in reply to https://github.com/AppPlatPerf/azure-cli/pull/60#discussion_r168366695

I had a strong opinion on this because of the issues we had in the past (e.g., we even had inconsistency issues with resource name length). It's a lot of extra work and user disappointment as well as effort on us to ask users to use a specific version of CLI (or clear their browser cache to load a newer version of portal).

I would suggest two other approaches here:

1. If it's a spec, move it to swagger to have a single source of truth that all SDKs/clients/RP (and third party apps if they care about consistency/correctness) can follow. CLI custom code should never be bothered with such a validation (especially considering the regex here is not a simple one).

2. Forget about validations in SDKs/clients and just use RP as a single source of truth. I would prefer this one since updating RP is much easier than updating swagger.

**Review by djyou — APPROVED** [review]

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 24 above)
Comment (sajayantony) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Lets just remove the word "Successfully" - Basically keep it as terse as possible. 
Also we need to indicate that you are zipping stuff up and how much we are uploading ..

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 6 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: updated. thanks

**Review by sajayantony — APPROVED** [review]

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 24 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Updated.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/sdk/models/azure-mgmt-containerregistry_enums.py (hunk 3 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/sdk/models/azure-mgmt-containerregistry_enums.py:46]: We will address it in next iteration.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 23 above)
Comment (northtyphoon) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py:243]: We will update it once we settle down the build configuration.
